gzochi-server/doc/gzochid.8
.TH GZOCHID 8 "April 1, 2016"
.SH NAME
.Nm gzochid - the gzochi server
.SH SYNOPSIS
\fBgzochid [\fR\fB\fIoptions\fR\fB]\fR
.SH DESCRIPTION
\fBgzochid\fR, the server for the gzochi massively multiplayer 
online game development framework, is the main program that does the
work of hosting deployed games and exposing endpoints for client
connections.
.P
When gzochid starts, it launches the games it finds in its 
deployment directory, storing their data in its data directory; it
begins listening for client connections and routing them to game
application endpoints; and optionally starts a web server that 
reports information about the state of the game server and any
active games.
.P
The \fBgzochid\fR program has many options that can be specified at
startup. For a complete list of options, run this command:
.IP
shell> \fBgzochid \-\-help\fR
.SH ENVIRONMENT
.TP
.B GZOCHID_CONF_LOCATION
If \fB$GZOCHID_CONF_LOCATION\fR is set before \fBgzochid\fR is started, its 
value is used as the full path to the server configuration file. This variable 
will be overriden by a command-line argument giving the location of the file, if
present.
.SH SEE ALSO
.P
The full documentation for \fBgzochid\fR is maintained as a Texinfo
manual. If the \fBinfo\fR and \fBgzochid\fR programs are properly 
installed at your site, the command
.IP
.B info gzochid
.P
should give you access to the complete manual.
.SH COPYRIGHT
Copyright \(co 2016 Julian Graham. License GPLv3+: GNU GPL version 3
or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH AUTHOR
Julian Graham (joolean@undecidable.net)


NEWS
gzochi NEWS -- history of user-visible changes
Copyright (C) 2015 Julian Graham
See the end of copying conditions.

Send comments, questions, and bug reports to gzochi-devel@nongnu.org

Changes in 0.8 (since 0.7)

* Changes to the distribution

** Removed deprecated synchronous client (libgzochi)

libgzochi, the synchronous reference client implementation included with earlier
versions of the gzochi framework, has been removed from the distribution. The
client was deprecated in an earlier release, since the concurrency strategy
suggested by its API was undesirable. Application developers are encouraged to
use libgzochi-glib instead.

** Removed client support static library (libgzochi-client-common)

libgzochi-client-common, the static library that facilitated sharing of code
between libgzochi and libgzochi-glib, has been merged with libgzochi-glib and no
longer exists as an independent build target within the gzochi framework source
distribution.

** New example: External transactions

Client and server code for a new example application has been added to the 
source distribution under the "examples/" tree. This example demonstrates 
techniques for synchronizing a transaction executed within the gzochid container
with an external transaction-aware process, such as a relational database.


Changes in 0.7 (since 0.6)

* Changes to the distribution

** New reference client library implementation: (gzochi client)

Game developers who want to keep their entire application stack in Guile Scheme
(C server container notwithstanding) are in luck! The new (gzochi client)
library is a Scheme reference client implementation with an execution model that
can be integrated into a main loop or used stand-alone. The source distribution
for (gzochi client) is under the "gzochi-guile-client/" tree.


Changes in 0.6 (since 0.5)

* Compatibility

** Support for additional platforms

gzochi now builds on Mac OS X and (likely) other BSD variants.


Changes in 0.3 (since 0.2)

* Changes to the distribution

** New client support static library: libgzochi-client-common

Some of the trickier aspects of client communication with a gzochi server
instance, such as maintaining a buffer for partial message delivery, have been
factored out of libgzochi and into a static library, with a source distribution
under the "gzochi-client-common/" tree, that can be built into other client 
implementations. 

** New reference client library implementation: libgzochi-glib

libgzochi-client-common already has a consumer in the distribution:
libgzochi-glib, a new reference client implementation that integrates easily 
with the "main loop" functionality in the popular GNU GLib library to support
the development of single-threaded client applications. The source distribution
for libgzochi-glib is under the "gzochi-glib-client/" tree.

** Removed deprecated debugging client

mictlan, the primitive "debugging client" included with earlier versions of the
gzochi framework, has been removed from the distribution. Its functionality has
been superceded by the debugging capabilities of gzochid itself, which can be 
accessed using a telnet client.


Changes in 0.2 (since 0.1)

* Changes to the distribution

** New example: abermud

Client and server code for a new example game -- based on Alan Cox's 
AberMUD -- has been added to the source distribution under the "examples/"
tree. This example demonstrates techniques for maintaining a persistent world
in which state is shared between multiple players.


Copyright information:

Copyright (C) 2015 Julian Graham

   Permission is granted to anyone to make or distribute verbatim copies
   of this document as received, in any medium, provided that the
   copyright notice and this permission notice are preserved,
   thus giving the recipient permission to redistribute in turn.

   Permission is granted to distribute modified versions
   of this document, or of portions of it,
   under the above conditions, provided also that they
   carry prominent notices stating who last changed them.


Local variables:
mode: outline
paragraph-separate: "[ 	]*$"
end:


gzochi-server/src/gzochid.c
/* gzochid.c: Main server bootstrapping routines for gzochid
 * Copyright (C) 2016 Julian Graham
 *
 * gzochi is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>
#include <getopt.h>
#include <glib.h>
#include <locale.h>
#include <libintl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>

#include "admin.h"
#include "config.h"
#include "context.h"
#include "fsm.h"
#include "game.h"
#include "guile.h"
#include "gzochid.h"
#include "log.h"
#include "threads.h"

#define _(String) gettext (String)

#define Q(x) #x
#define QUOTE(x) Q(x)

#ifndef GZOCHID_CONF_LOCATION
#define GZOCHID_CONF_LOCATION "/etc/gzochid.conf"
#endif /* GZOCHID_CONF_LOCATION */

static const struct option longopts[] =
  {
    { "config", required_argument, NULL, 'c' },
    { "help", no_argument, NULL, 'h' },
    { "version", no_argument, NULL, 'v' },
    { NULL, 0, NULL, 0 }
  };

static void 
initialize_async (gpointer data, gpointer user_data)
{
  gzochid_context *context = user_data;
  gzochid_server_context *server_context = (gzochid_server_context *) context;

  GKeyFile *key_file = g_key_file_new ();
  GHashTable *admin_config = NULL;
  GHashTable *game_config = NULL;  
  GHashTable *log_config = NULL;

  GError *err = NULL;

  server_context->admin_context = 
    (gzochid_context *) gzochid_admin_context_new ();
  server_context->game_context =
    (gzochid_context *) gzochid_game_context_new ();

  g_key_file_load_from_file 
    (key_file, server_context->gzochid_conf_path, G_KEY_FILE_NONE, &err);

  if (err != NULL)
    {
      /* 
	 The server has reasonable defaults for everything in gzochid.conf, so
	 this condition is not fatal.
      */

      g_warning 
	("Failed to load server configuration file %s: %s", 
	 server_context->gzochid_conf_path, err->message);
      g_error_free (err);
    }

  admin_config = gzochid_config_keyfile_extract_config (key_file, "admin");
  game_config = gzochid_config_keyfile_extract_config (key_file, "game");
  log_config = gzochid_config_keyfile_extract_config (key_file, "log");

  if (g_hash_table_contains (log_config, "priority.threshold")) 
    {
      char *threshold = g_hash_table_lookup (log_config, "priority.threshold");

      if (g_ascii_strncasecmp (threshold, "EMERG", 5) == 0)
	gzochid_install_log_handler (G_LOG_LEVEL_ERROR);
      else if (g_ascii_strncasecmp (threshold, "ALERT", 5) == 0)
	gzochid_install_log_handler (G_LOG_LEVEL_ERROR);
      else if (g_ascii_strncasecmp (threshold, "CRIT", 4) == 0)
	gzochid_install_log_handler (G_LOG_LEVEL_ERROR);
      else if (g_ascii_strncasecmp (threshold, "ERR", 3) == 0)
	gzochid_install_log_handler (G_LOG_LEVEL_CRITICAL);
      else if (g_ascii_strncasecmp (threshold, "WARNING", 7) == 0)
	gzochid_install_log_handler (G_LOG_LEVEL_WARNING);
      else if (g_ascii_strncasecmp (threshold, "NOTICE", 6) == 0)
	gzochid_install_log_handler (G_LOG_LEVEL_MESSAGE);
      else if (g_ascii_strncasecmp (threshold, "INFO", 4) == 0)
	gzochid_install_log_handler (G_LOG_LEVEL_INFO);
      else if (g_ascii_strncasecmp (threshold, "DEBUG", 5) == 0)
	gzochid_install_log_handler (G_LOG_LEVEL_DEBUG);
    }
  else gzochid_install_log_handler (G_LOG_LEVEL_INFO);

  gzochid_admin_context_init 
    ((gzochid_admin_context *) server_context->admin_context, context, 
     admin_config);
  gzochid_game_context_init 
    ((gzochid_game_context *) server_context->game_context, context, 
     game_config);

  gzochid_context_until 
    (server_context->admin_context, GZOCHID_ADMIN_STATE_RUNNING);
  gzochid_context_until 
    (server_context->game_context, GZOCHID_GAME_STATE_RUNNING);

  g_hash_table_unref (admin_config);
  g_hash_table_unref (game_config);
  g_hash_table_unref (log_config);

  g_key_file_free (key_file);

  gzochid_fsm_to_state (context->fsm, GZOCHID_STATE_RUNNING);
}

static void 
initialize (int from_state, int to_state, gpointer user_data)
{
  gzochid_server_context *context = user_data;
  gzochid_thread_pool_push (context->pool, initialize_async, NULL, NULL);
}

static void *
initialize_guile_inner (void *ptr)
{
  gzochid_guile_init ();
  return NULL;
}

static void 
initialize_guile (int from_state, int to_state, gpointer user_data)
{
  scm_with_guile (initialize_guile_inner, NULL);
}

gzochid_server_context *
gzochid_server_context_new (void)
{
  return calloc (1, sizeof (gzochid_server_context));
}

void 
gzochid_server_context_init (gzochid_server_context *context)
{
  gzochid_fsm *fsm = gzochid_fsm_new 
    ("main", GZOCHID_STATE_INITIALIZING, "INITIALIZING");

  gzochid_fsm_add_state (fsm, GZOCHID_STATE_RUNNING, "RUNNING");
  gzochid_fsm_add_state (fsm, GZOCHID_STATE_STOPPED, "STOPPED");

  context->pool = gzochid_thread_pool_new (context, 1, TRUE, NULL);

  gzochid_fsm_on_enter 
    (fsm, GZOCHID_STATE_INITIALIZING, initialize_guile, context);
  gzochid_fsm_on_enter 
    (fsm, GZOCHID_STATE_INITIALIZING, initialize, context);
  
  gzochid_fsm_add_transition 
    (fsm, GZOCHID_STATE_INITIALIZING, GZOCHID_STATE_RUNNING);
  gzochid_fsm_add_transition 
    (fsm, GZOCHID_STATE_RUNNING, GZOCHID_STATE_STOPPED);

  gzochid_context_init ((gzochid_context *) context, NULL, fsm);
}

void 
gzochid_server_context_free (gzochid_server_context *context)
{
  gzochid_context *root_context = (gzochid_context *) context;
  gzochid_fsm_free (root_context->fsm);

  free (context);
}

static void
print_help (const char *program_name)
{
  printf (_("\
Usage: %s [OPTION]...\n"), program_name);
  
  puts ("");
  fputs (_("\
  -c, --config        full path to gzochid.conf\n\
  -h, --help          display this help and exit\n\
  -v, --version       display version information and exit\n"), stdout);

  puts ("");
  printf (_("\
Report bugs to: %s\n"), PACKAGE_BUGREPORT);
#ifdef PACKAGE_PACKAGER_BUG_REPORTS
  printf (_("Report %s bugs to: %s\n"), PACKAGE_PACKAGER,
          PACKAGE_PACKAGER_BUG_REPORTS);
#endif /* PACKAGE_PACKAGER_BUG_REPORTS */

#ifdef PACKAGE_URL
  printf (_("%s home page: <%s>\n"), PACKAGE_NAME, PACKAGE_URL);
#else
  printf (_("%s home page: <http://www.nongnu.org/%s/>\n"),
          PACKAGE_NAME, PACKAGE);
#endif /* PACKAGE_URL */
}

static void
print_version (void)
{
  printf ("gzochid (gzochi) %s\n", VERSION);

  puts ("");
  printf (_("\
Copyright (C) %s Julian Graham\n\
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n\
This is free software: you are free to change and redistribute it.\n\
There is NO WARRANTY, to the extent permitted by law.\n"),
	  "2015");
}

int 
main (int argc, char *argv[])
{
  gzochid_server_context *context = NULL;
  const char *program_name = argv[0];
  int optc = 0;

  setlocale (LC_ALL, "");

  context = gzochid_server_context_new ();

  while ((optc = getopt_long (argc, argv, "c:hv", longopts, NULL)) != -1)
    switch (optc)
      {
      case 'c':
	context->gzochid_conf_path = strdup (optarg);
	break;
      case 'v':
	print_version ();
	exit (EXIT_SUCCESS);
	break;
      case 'h':
	print_help (program_name);
	exit (EXIT_SUCCESS);
	break;
      case '?': exit (EXIT_FAILURE);
      }

  if (context->gzochid_conf_path == NULL)
    {
      const char *env = getenv ("GZOCHID_CONF_LOCATION");
      context->gzochid_conf_path = env ? env : QUOTE(GZOCHID_CONF_LOCATION);
    }

  g_info ("Reading configuration from %s", context->gzochid_conf_path);
  
  gzochid_server_context_init (context);
  gzochid_fsm_until (((gzochid_context *) context)->fsm, GZOCHID_STATE_STOPPED);
  gzochid_server_context_free (context);
	   
  return 0;
}
